Fix TransactionRunner, revert changes made in earlier commit

dlmarion opened the pull request.

milleruntime commented:
Since this is difficult to review, I checked out your branch and did a diff with what was in Fate.java before your changes:
```diff --git a/core/src/main/java/org/apache/accumulo/fate/Fate.java b/core/src/main/java/org/apache/accumulo/fate/Fate.java
index cf02c56cf6..cd53546b74 100644
--- a/core/src/main/java/org/apache/accumulo/fate/Fate.java
+++ b/core/src/main/java/org/apache/accumulo/fate/Fate.java
@@ -21,6 +21,9 @@ package org.apache.accumulo.fate;
 import java.io.IOException;
 import java.util.EnumSet;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Function;
@@ -50,6 +53,7 @@ public class Fate<T> {
 
   private final TStore<T> store;
   private final T environment;
+  private ScheduledThreadPoolExecutor fatePoolWatcher;
   private ExecutorService executor;
 
   private static final EnumSet<TStatus> FINISHED_STATES =
@@ -113,7 +117,6 @@ public class Fate<T> {
             store.unreserve(tid, deferTime);
           }
         }
-
       }
     }
 
@@ -226,11 +229,33 @@ public class Fate<T> {
    * Launches the specified number of worker threads.
    */
   public void startTransactionRunners(AccumuloConfiguration conf) {
-    int numThreads = conf.getCount(Property.MANAGER_FATE_THREADPOOL_SIZE);
-    executor = ThreadPools.createExecutorService(conf, Property.MANAGER_FATE_THREADPOOL_SIZE);
-    for (int i = 0; i < numThreads; i++) {
-      executor.execute(new TransactionRunner());
-    }
+    final ThreadPoolExecutor pool = (ThreadPoolExecutor) ThreadPools.createExecutorService(conf,
+        Property.MANAGER_FATE_THREADPOOL_SIZE);
+    fatePoolWatcher = ThreadPools.createGeneralScheduledExecutorService(conf);
+    fatePoolWatcher.schedule(() -> {
+      // resize the pool if the property changed
+      ThreadPools.resizePool(pool, conf, Property.MANAGER_FATE_THREADPOOL_SIZE);
+      // If the pool grew, then ensure that there is a TransactionRunner for each thread
+      int needed = conf.getCount(Property.MANAGER_FATE_THREADPOOL_SIZE) - pool.getQueue().size();
+      if (needed > 0) {
+        for (int i = 0; i < needed; i++) {
+          try {
+            pool.execute(new TransactionRunner());
+          } catch (RejectedExecutionException e) {
+            // RejectedExecutionException could be shutting down
+            if (pool.isShutdown()) {
+              // The exception is expected in this case, no need to spam the logs.
+              log.trace("Error adding transaction runner to FaTE executor pool.", e);
+            } else {
+              // This is bad, FaTE may no longer work!
+              log.error("Error adding transaction runner to FaTE executor pool.", e);
+            }
+            break;
+          }
+        }
+      }
+    }, 3, TimeUnit.SECONDS);
+    executor = pool;
   }
 
   // get a transaction id back to the requester before doing any work
@@ -325,6 +350,7 @@ public class Fate<T> {
    */
   public void shutdown() {
     keepRunning.set(false);
+    fatePoolWatcher.shutdown();
     executor.shutdown();
   }
```

I assume this is a follow-up fix for your fix in #2278 but it's hard to tell without a description or link to the earlier commit. If so, I think this revert is OK. What do you think, do these changes look correct?

Manno15 reviewed (approved):
ITs pass now with these changes. The diff looks as expected if the goal was just to revert the run() part.

milleruntime commented on core/src/main/java/org/apache/accumulo/fate/Fate.java:
```diff
@@ -228,24 +233,25 @@ public void startTransactionRunners(AccumuloConfiguration conf) {
-            // This is bad, FaTE may no longer work!
-            log.error("Error adding transaction runner to FaTE executor pool.", e);
+      // If the pool grew, then ensure that there is a TransactionRunner for each thread
+      int needed = conf.getCount(Property.MANAGER_FATE_THREADPOOL_SIZE) - pool.getQueue().size();
+      if (needed > 0) {
```
It also looks this way doesn't allow for reducing the size of the threadpool.

dlmarion replied:
The call to `ThreadPools.resizePool` will either increase or decrease the number of Threads in the pool depending on the current value of the Property.

milleruntime replied:
Ah ok I see now.

dlmarion commented:
I think you are seeing different issues being addressed.These changes do look correct, here's why.

1. [Here](https://github.com/apache/accumulo/commit/e62ace6a9d37572d95999f5412c6148efbba50b9#diff-467b6cb7af0f7b7339d81d3112329b8e19557e685a3c659bc623160826c8f01dL228) I modified Fate.java to use the new ThreadPools class as part of ThreadPools PR.

2. [Here](https://github.com/apache/accumulo/commit/e67ffa0a4be2140234fb3c0ea0742dff192bce91#diff-467b6cb7af0f7b7339d81d3112329b8e19557e685a3c659bc623160826c8f01dL229) includes changes for multiple issues. The change is part of a PR that closes #2272 and #2276. #2272 was an issue that @keith-turner submitted where he noticed that properties marked as mutable were not being honored. Specifically, there were some thread pool properties that were marked as mutable but were either not being resized or were done differently. I consolidated the pool resize logic and updated the code to use it for the ThreadPools that were marked as mutable in a PR for #2278.

3. As part of #2278 I also included a resolution to [this](https://github.com/apache/accumulo/issues/2272#issuecomment-922119524) comment in #2272 . This is where the bug lies.

This PR reverts the code changes for # 3 above, keeps the ThreadPool resizing, and ensures that there is a TransactionRunner running for each Thread.

milleruntime commented:
@dlmarion thanks for the detailed explanation, that helps a lot.

milleruntime commented:
+1 for the branch name BTW
